### PR TITLE
Add parser and some semantic support for gdc extended asm statements

### PR DESCRIPTION
**Pull request by yebblies**
@ibuclaw @klickverbot GDC is the only one that uses this, right?  This is a large chunk of the remaining differences in #2194

### CODE REVIEW

**Comment by ibuclaw** [comment]
@yebblies - did you pull this straight from GDC?

Not sure if there are a couple of places missing semantic-wise.  I'll have a check.

**Comment by yebblies** [comment]
> did you pull this straight from GDC?

Yep!  With a couple of minor changes.  I pretty much searched for `ExtAsmStatement` and copied everything that looked relevant.

**Comment by yebblies** [comment]
Maybe I should change the error message then, so you can use it as-is.

**Comment by WalterBright** [comment]
I don't know what this does.

**Comment by yebblies** [comment]
> I don't know what this does.

This is one of the outstanding patches that GDC has to apply to the frontend source.  Ideally no patches need to be applied to the frontend at all, so that most compiler updates can be automatically applied to GDC and LDC.  This adds parsing support to dmd for the GDC-only extended asm statements.

> Also, if this is "pulled straight from GDC" is it GPL code? If so, we can't use it in DMD.

It's part of the frontend, so I _think_ it's available or can be made available under boost.  @ibuclaw ?

**Comment by ibuclaw** [comment]
@yebblies @WalterBright - It can be made available under boost.  Otherwise I would have added a license exception at the top of the file.

**Comment by WalterBright** [comment]
> This adds parsing support to dmd for the GDC-only extended asm statements.

1. What are they? (link to documentation?)
2. Would that grammar then work in dmd?
3. Should it be GDC-only?

**Comment by ibuclaw** [comment]
1. http://wiki.dlang.org/GDC/Using_GDC#Extended_Assembler
GCC has better documentation though: https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html

The main differences between GCC and GDC extended assembler is noted in the GDC wiki.

2. Yes

3. Yes - LDC as a similarly difference syntax (works more like a function call).  http://wiki.dlang.org/LDC_inline_assembly_expressions

GDC extended assembler can be thought of as a kind of asm_printf() which takes the given format string, and writes it out in the resultant assembly (.S) that the compiler emits.  This of course won't work for DMD because it emits object code (.o) only.

File: src/interpret.c [hunk 1]
```diff
@@ -683,6 +683,14 @@ class CtfeCompiler : public Visitor
         // we can't compile asm statements
     }
 
+    void visit(ExtAsmStatement *s)
+    {
+    #if LOGCOMPILE
+        printf("%s AsmStatement::ctfeCompile\n", s->loc.toChars());
```
Comment (ibuclaw) [inline src/interpret.c]: ``ExtAsmStatement::ctfeCompile``

Comment (yebblies) [inline src/interpret.c]: Fixed.

File: src/statement.c [hunk 2]
```diff
@@ -5060,6 +5061,101 @@ Statement *AsmStatement::syntaxCopy()
...
+    if (args)
+    {
+        for (size_t i = 0; i < args->dim; i++)
+        {
+            Expression *e = (*args)[i];
+            e = e->semantic(sc);
+            if (i < outputargs)
+                e = e->modifiableLvalue(sc, NULL);
+            else
+                e = e->optimize(WANTvalue);
```
Comment (ibuclaw) [inline src/statement.c]: This is a mistake here (there's no check for a valid rvalue).  Should be:
```
e = e->optimize(WANTvalue);
// Check argument is a valid lvalue/rvalue.
if (i < outputargs)
    e = e->modifiableLvalue(sc, NULL);
else
    e->rvalue();
```

Comment (yebblies) [inline src/statement.c]: Ok.

File: src/statement.c [hunk 3]
```diff
@@ -5060,6 +5061,101 @@ Statement *AsmStatement::syntaxCopy()
...
+    // Fold all clobbers.
+    if (clobbers)
+    {
+        for (size_t i = 0; i < clobbers->dim; i++)
+        {
+            Expression *e = (*clobbers)[i];
+            e = e->semantic(sc);
+            e = e->optimize(WANTvalue);
+            if (e->op != TOKstring || ((StringExp *) e)->sz != 1)
+                error("clobber specification must be a constant char string");
```
Comment (ibuclaw) [inline src/statement.c]: Should be ``e->error`` for accurate column highlighting.

File: src/statement.c [hunk 4]
```diff
@@ -5060,6 +5061,101 @@ Statement *AsmStatement::syntaxCopy()
...
+                e = e->modifiableLvalue(sc, NULL);
+            else
+                e = e->optimize(WANTvalue);
+            (*args)[i] = e;
+
+            e = (*constraints)[i];
+            e = e->semantic(sc);
+            e = e->optimize(WANTvalue);
+            if (e->op != TOKstring || ((StringExp *) e)->sz != 1)
+                error ("constraint must be a constant char string");
```
Comment (ibuclaw) [inline src/statement.c]: Should be ``e->error`` for accurate column highlighting

File: src/statement.c [hunk 5]
```diff
@@ -5060,6 +5061,101 @@ Statement *AsmStatement::syntaxCopy()
...
+}
+
+Statement *ExtAsmStatement::semantic(Scope *sc)
+{
+    // Fold the instruction template string.
+    insn = insn->semantic(sc);
+    insn->optimize(WANTvalue);
+
+    if (insn->op != TOKstring || ((StringExp *) insn)->sz != 1)
+        error("instruction template must be a constant char string");
```
Comment (ibuclaw) [inline src/statement.c]: Should be ``insn->error``

File: src/statement.c (hunk 4 above)
Comment (ibuclaw) [inline src/statement.c]: Actually, this is enforced by the parser, should probably make it as assert then.

File: src/statement.c (hunk 3 above)
Comment (ibuclaw) [inline src/statement.c]: Actually, this is enforced by the parser, should probably make it an assert then.

**Comment by yebblies** [comment]
@ibuclaw Probably easier if you give me a link to the diff when you fix these in gdc.

File: src/statement.c (hunk 5 above)
Comment (ibuclaw) [inline src/statement.c]: Also need an extra check below this:
```
if (labels && outputargs)
    error("extended asm statements with labels cannot have output constraints");
```
I'll send you a PR for all I've mentioned. ;)

**Comment by ibuclaw** [comment]
@yebblies - https://github.com/D-Programming-GDC/GDC/commit/c76a734d3bf6db8120d4262fd2bba53c58174e29

**Comment by braddr** [comment]
I know that the inline asm syntax is an area where we might not have any choices on being different between the compilers (though I'm not sold on that, but I do agree that the amount of work might not be practical), I'm very leery about endorsing that difference by having a different version for each compiler in the official front end source.  This feels like an area that should remain in the compiler specific patches.

**Comment by ibuclaw** [comment]
@braddr - Maybe if AsmStatement were transparent to the D frontend, and the parsing be left up to compiler-specific interpretation?

**Comment by yebblies** [comment]
> This feels like an area that should remain in the compiler specific patches.

It'll be hard to reach zero patches that way.

**Comment by ibuclaw** [comment]
Moved to  #7562.

**Pull request closed without merging**